src/components/SearchBar.ts

```diff
@@ -0,0 +1,127 @@
+                    ref: input => this.searchInput = input as HTMLInputElement
+                }),
+                createElement("button", {
+                    className: `btn-transparent ${this.state.buttonVisibility}`,
+                    ref: button => this.searchButton = button as HTMLButtonElement
+                },
+                    createElement("span", { className: "glyphicon glyphicon-remove" })
+                )
+            )
+            : createElement("div", { className: "search-container-hidden" });
+    }
+
+    componentWillReceiveProps(nextProps: SearchBarProps) {
+        if (this.props.showSearchBar && this.props.listView) {
+            this.setUpEvents();
+            this.setDefaultSearch(this.props.defaultQuery);
+        }
+    }
+
+    componentWillUnmount() {
+        if (this.props.showSearchBar && this.props.listView) {
+            this.searchButton.removeEventListener("click", this.onClear);
+            this.searchInput.removeEventListener("keyup", this.onSearchKeyDown);
+        }
+    }
+
+    private setUpEvents() {
+        this.searchButton.addEventListener("click", this.onClear);
+        this.searchInput.addEventListener("keyup", this.onSearchKeyDown);
+    }
+
+    private setDefaultSearch(query: string) {
+        this.searchInput.value = query;
+        this.updateConstraints(this);
+    }
+
+    private onSearchKeyDown(event: CustomEvent) {
+        this.updateButtonVisibility();
+        const searchTimeout = setTimeout(this.updateConstraints(this), 500);
+        clearTimeout(searchTimeout);
+    }
+
+    private updateButtonVisibility() {
+        if (this.searchInput.value.trim()) {
+            this.setState({ buttonVisibility: "visible" });
+        } else {
+            this.setState({ buttonVisibility: "hidden" });
+        }
+    }
+
+    private updateConstraints(self: SearchBar) {
+        if (self.props.listView) {
+            const datasource = self.props.listView._datasource;
+            if (window.device) {
+                const constraints: HybridConstraint = [];
+                self.props.searchAttributes.forEach(searchAttribute => {
+                    if (this.props.searchEntity) {
+                        constraints.push({
+                            attribute: searchAttribute.name,
+                            operator: self.props.searchMethod,
+                            path: self.props.searchEntity,
+                            value: self.searchInput.value
+                        });
+                    } else {
+                        constraints.push({
+                            attribute: searchAttribute.name,
+                            operator: self.props.searchMethod,
+                            value: self.searchInput.value
+                        });
+                    }
+                });
+                self.searchInput.value.trim() ? datasource._constraints = constraints : datasource._constraints = [];
```

Comment:
Don`t like this. Ternary operators should be uses for assigning a value to return, assiging them in the true or false part!
```
datasource._constraints = self.searchInput.value.trim() ?  constraints : [];
```